src/libiotc/iotc_data_desc.c

```diff
@@ -92,11 +92,11 @@ iotc_data_desc_t* iotc_make_desc_from_string_copy(const char* str) {
   IOTC_ALLOC(iotc_data_desc_t, data_desc, state);
   data_desc->memory_type = IOTC_MEMORY_TYPE_MANAGED;
 
-  IOTC_ALLOC_BUFFER_AT(unsigned char, data_desc->data_ptr, len, state);
-  memcpy(data_desc->data_ptr, str, len);
+  IOTC_ALLOC_BUFFER_AT(unsigned char, data_desc->data_ptr, len + 1, state);
```

Comment:
I'd rather not include this change in the fix. Why? For these reasons:
- first I'd not touch the design of the `iotc_data_desc_t`. By design the data and its size is specified with two members: [`data_ptr` and the corresponding `length`](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/1c1bdcb12d408a7e93d3873824088e97c6283c8a/src/libiotc/iotc_data_desc.h#L32-L39).
- including the string terminator null is redundant. It is one extra byte which isn't necessary. One might say one extra byte won't harm anybody, but this library is designed to use as small memory as possible.
- what is the situation with [the usage place](https://github.com/GoogleCloudPlatform/iot-device-sdk-embedded-c/blob/master/src/libiotc/iotc.c#L544)? It will then publish a message with the null terminator included. Although - another example for other length indicator - the [MQTT protocol does not use null terminator](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/errata01/os/mqtt-v3.1.1-errata01-os-complete.html#_Toc442180853) to indicate end of a string/PUBLISH payload.

So above there are 3 different types of string length indicators, null terminator is just one of them. I'd rather adhere to the current design and use the `iotc_data_desc_t` length representation. Additionally we should then alter the `iotc_make_desc_from_string_share` function as well to use `strlen(str) + 1` as `length`. And maybe other places should be tracked down to switch to the new 'policy': if string is stored in `iotc_data_desc_t`, then indicate the end with null. Then next question arises, how to remember what type of data is stored in this `iotc_data_desc_t` instance? Should we then add an enum member to it? And even then if binary is stored (it may contain null inside) there is a chance to make a mistake and read it as a string. So I'd keep it simple, one length representation, adhering to current `iotc_data_desc_t` design.

Reply:
Reverted. Thanks for the input.